bundles/org.openhab.binding.pentair/src/main/java/org/openhab/binding/pentair/internal/handler/PentairBaseBridgeHandler.java

```diff
@@ -99,355 +143,317 @@ public void handleCommand(ChannelUID channelUID, Command command) {
-                disconnect();
-                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
-            } catch (EOBException e) {
-                // EOB Exception is used to exit the parser loop if full message is not in buffer.
+
+                logger.debug("First command from unregistered device ({}): {}", p.getSource(), p);
+                unregistered.add(p.getSource());
+            } else {
+                logger.trace("Subsequent command from unregistered device ({}): {}", p.getSource(), p);
             }
+        } else {
+            logger.debug("Received pentair command: {}", p);
 
-            logger.debug("msg reader thread exited");
+            thinghandler.processPacketFrom(p);
+            ackResponse(p.getAction());
         }
     }
 
+    @Override
+    public void onIntelliChlorPacket(PentairIntelliChlorPacket p) {
+        PentairBaseThingHandler thinghandler;
+
+        thinghandler = equipment.get(0);
+
+        if (thinghandler == null) {
+            if (!unregistered.contains(0)) { // if not yet seen, print out log message
+                if (discovery && discoveryService != null) {
+                    discoveryService.notifyDiscoveredIntellichlor(0);
+                }
+                logger.debug(" First command from unregistered Intelliflow: {}", p);
+                unregistered.add(0);
+            } else {
+                logger.trace("Subsequent command from unregistered Intelliflow: {}", p);
+            }
+
+            return;
+        }
+
+        thinghandler.processPacketFrom(p);
+    }
+
     /**
      * Method to write a package on the Pentair bus. Will add preamble and checksum to bytes written
      *
      * @param p {@link PentairPacket} to write
      */
     public void writePacket(PentairPacket p) {
-        try { // FF 00 FF A5 00 60 10 07 00 01 1C
-            byte[] preamble = { (byte) 0xFF, (byte) 0xFF, (byte) 0xFF, (byte) 0x00, (byte) 0xFF };
-            byte[] buf = new byte[5 + p.getLength() + 8]; // 5 is preamble, 8 is 6 bytes for header and 2 for checksum
+        writePacket(p, -1, 0);
+    }
+
+    public boolean writePacket(PentairPacket p, int response, int retries) {
+        boolean bReturn = true;
 
+        try {
+            byte[] buf;
+            int nRetries = retries;
+
+            if (!writer.isPresent()) {
+                logger.debug("writePacket: writer = null");
+                return false;
+            }
             p.setSource(id);
 
-            System.arraycopy(preamble, 0, buf, 0, 5);
-            System.arraycopy(p.buf, 0, buf, 5, p.getLength() + 6);
-            int checksum = p.calcChecksum();
+            buf = p.getFullWriteStream();
 
-            buf[p.getLength() + 11] = (byte) ((checksum >> 8) & 0xFF);
-            buf[p.getLength() + 12] = (byte) (checksum & 0xFF);
+            lock.lock();
+            ackResponse = response;
```

Comment:
Locking should be done as either immediately before, or immediately after the `try {` statement.

I think the code above the `lock.lock()` could be moved above the `try {` statement.